doc/vpectrl.8
.\" Automatically generated by Pod::Man v1.36, Pod::Parser v1.13
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` 
.    ds C' 
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "VPECTRL 8"
.TH VPECTRL 8 "2003-03-28" "0.1" "Virtual Private Ethernet"
.SH "NAME"
\&\f(CW\*(C`vpectrl\*(C'\fR \- Virtual Private Ethernet Control Program
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\f(CW\*(C`vpectrl\*(C'\fR [\fB\-ckgs\fR] [\fB\-\-config=\fR\fI\s-1DIR\s0\fR] [\fB\-\-generate\-keys\fR] [\fB\-\-help\fR]
[\fB\-\-kill\fR[\fB=\fR\fI\s-1SIGNAL\s0\fR]] [\fB\-\-show\-config\fR] [\fB\-\-version\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This is the control program for the \f(CW\*(C`vped\*(C'\fR, the virtual private ethernet daemon.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-c\fR, \fB\-\-config=\fR\fI\s-1DIR\s0\fR" 4
.IX Item "-c, --config=DIR"
Read configuration options from \fI\s-1DIR\s0\fR.
.IP "\fB\-g\fR, \fB\-\-generate\-keys\fR" 4
.IX Item "-g, --generate-keys"
Generate public/private \s-1RSA\s0 keypair and exit.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
Display short list of options.
.IP "\fB\-\-kill\fR[\fB=\fR\fI\s-1SIGNAL\s0\fR]" 4
.IX Item "--kill[=SIGNAL]"
Attempt to kill a running \f(CW\*(C`vpectrl\*(C'\fR (optionally with the specified
\&\fI\s-1SIGNAL\s0\fR instead of \f(CW\*(C`SIGTERM\*(C'\fR) and exit.
.IP "\fB\-\-show\-config\fR" 4
.IX Item "--show-config"
Show a summary of the configuration, and how vped interprets it. Can also be very useful
when designing firewall scripts.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
Output version information and exit.
.SH "BUGS"
.IX Header "BUGS"
If you find any bugs, report them to \f(CW\*(C`vpe@plan9.de\*(C'\fR.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIvpe\fR\|(5), \fIvpe.conf\fR\|(5), \fIvped\fR\|(8).
.PP
vpe comes with \s-1ABSOLUTELY\s0 \s-1NO\s0 \s-1WARRANTY\s0.  This is free software, and you are
welcome to redistribute it under certain conditions; see the file \s-1COPYING\s0
for details.
.SH "AUTHORS"
.IX Header "AUTHORS"
Marc Lehmann \f(CW\*(C`<vpe@plan9.de>\*(C'\fR.


NEWS
1.6 Mon May 10 20:55:10 CEST 2004
   - protocol version 0.1, compatible with older releases.
   - do not RESET on out-of-sequence packets (good for wireless).
   - various non-security-related bugfixes.
   - c++ify (at least make it compile with g++-3.4, which should make
     it a little bit more standard c++).

1.5 Fri Jan 30 00:50:04 CET 2004
   - protocol version 0.1, compatible with older releases.
   - vped will refuse to start when hostkey and public key do not match.
   - updated lzf code to version 1.2.
   - better error reporting for "unusual" conditions, like failing
     to allocate memory, that should not normally happen and
     will otherwise result in spurious other error messages. Also
     adds paranoid checks in case openssl's API changes in a bad way.
   - fix a bug where queued vpn packets were cleared to zero. while
     this doesn't affect security, it did cause warning messages and
     unnecessary connectivity delays.

1.4 Sat Jan 17 15:49:21 CET 2004
   - protocol version 0.1, compatible with older releases.
   - better retry behaviour on key mismatch.
   - use select-based io_manager instead of poll-based one.
   - share io manager between rxvt-unicode and vpe.
   - sooo many *BSD workarounds because no BSD I could find is even
     remotely POSIX-compatible.

1.2 Fri Oct 17 03:44:44 CEST 2003
   - protocol version 0.1.
   - tincd kernel interface code imported, more supported platforms
     (native/linux (2.4), tincd/linux (2.2 and 2.4), tincd/freebsd,
     native/cygwin).
   - added primitive ethernet emulation (ipv4 only), which allows
     the following platforms that only have tun drivers to work:
     /* none yet tested */
   - portability workarounds, especially for unfriendly freebsd
   - very minor bugfixes
   - warnings when choosing insecure ciphers/hashes
   - reduced default hmac length to 8.
   - cvs now contains configure, Makefile.in and other generated files.
   - added doc/complex-example to the distribution.

1.0
   - protocol version 0.1.
   - tweaked various timeouts to help very slow (486) class
     machines or nets with many hosts.
   - tweaked rate-limiting to be more forgiving for hosts
     connecting through routers (not a fix).



src/vpectrl.C
/*
    vpectrl.C -- the main file for vpectrl
    Copyright (C) 1998-2002 Ivo Timmermans <ivo@o2w.nl>
                  2000-2002 Guus Sliepen <guus@sliepen.eu.org>
                  2003      Marc Lehmannn <pcg@goof.com>
 
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
 
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 
    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc. 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include "config.h"

#include <cstdio>
#include <cstring>

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <signal.h>
#include <sys/types.h>
#include <unistd.h>
#include <signal.h>

#include <openssl/rand.h>
#include <openssl/rsa.h>
#include <openssl/pem.h>
#include <openssl/evp.h>

#include "pidfile.h"

#include "gettext.h"

#include "conf.h"
#include "slog.h"
#include "util.h"
#include "vpn.h"

/* If nonzero, display usage information and exit. */
static int show_help;

/* If nonzero, print the version on standard output and exit.  */
static int show_version;

/* If nonzero, it will attempt to kill a running vped and exit. */
static int kill_vped;

/* If nonzero, it will attempt to kill a running vped and exit. */
static int show_config;

/* If nonzero, generate public/private keypair for this net. */
static int generate_keys;

static struct option const long_options[] =
    {
      {"config", required_argument, NULL, 'c'},
      {"kill", optional_argument, NULL, 'k'},
      {"help", no_argument, &show_help, 1},
      {"version", no_argument, &show_version, 1},
      {"generate-keys", no_argument, NULL, 'g'},
      {"show-config", no_argument, &show_config, 's'},
      {NULL, 0, NULL, 0}
    };

static void
usage (int status)
{
  if (status != 0)
    fprintf (stderr, _("Try `%s --help\' for more information.\n"), get_identity ());
  else
    {
      printf (_("Usage: %s [option]...\n\n"), get_identity ());
      printf (_
              ("  -c, --config=DIR           Read configuration options from DIR.\n"
               "  -k, --kill[=SIGNAL]        Attempt to kill a running vped and exit.\n"
               "  -g, --generate-keys        Generate public/private RSA keypair.\n"
               "  -s, --show-config          Display the configuration information.\n"
               "      --help                 Display this help and exit.\n"
               "      --version              Output version information and exit.\n\n"));
      printf (_("Report bugs to <vpe@plan9.de>.\n"));
    }

  exit (status);
}

void
parse_options (int argc, char **argv, char **envp)
{
  int r;
  int option_index = 0;

  while ((r =
            getopt_long (argc, argv, "c:k::gs", long_options,
                         &option_index)) != EOF)
    {
      switch (r)
        {
        case 0:		/* long option */
          break;

        case 'c':		/* config file */
          confbase = strdup (optarg);
          break;

        case 'k':		/* kill old vpeds */
          if (optarg)
            {
              if (!strcasecmp (optarg, "HUP"))
                kill_vped = SIGHUP;
              else if (!strcasecmp (optarg, "TERM"))
                kill_vped = SIGTERM;
              else if (!strcasecmp (optarg, "KILL"))
                kill_vped = SIGKILL;
              else if (!strcasecmp (optarg, "USR1"))
                kill_vped = SIGUSR1;
              else if (!strcasecmp (optarg, "USR2"))
                kill_vped = SIGUSR2;
              else if (!strcasecmp (optarg, "WINCH"))
                kill_vped = SIGWINCH;
              else if (!strcasecmp (optarg, "INT"))
                kill_vped = SIGINT;
              else if (!strcasecmp (optarg, "ALRM"))
                kill_vped = SIGALRM;
              else
                {
                  kill_vped = atoi (optarg);

                  if (!kill_vped)
                    {
                      fprintf (stderr,
                               _
                               ("Invalid argument `%s'; SIGNAL must be a number or one of HUP, TERM, KILL, USR1, USR2, WINCH, INT or ALRM.\n"),
                               optarg);
                      usage (1);
                    }
                }
            }
          else
            kill_vped = SIGTERM;

          break;

        case 'g':		/* generate public/private keypair */
          generate_keys = RSA_KEYBITS;
          break;

        case 's':
          show_config = 1;
          break;

        case '?':
          usage (1);

        default:
          break;
        }
    }
}

/* This function prettyprints the key generation process */

void
indicator (int a, int b, void *p)
{
  switch (a)
    {
    case 0:
      fprintf (stderr, ".");
      break;

    case 1:
      fprintf (stderr, "+");
      break;

    case 2:
      fprintf (stderr, "-");
      break;

    case 3:
      switch (b)
        {
        case 0:
          fprintf (stderr, " p\n");
          break;

        case 1:
          fprintf (stderr, " q\n");
          break;

        default:
          fprintf (stderr, "?");
        }
      break;

    default:
      fprintf (stderr, "?");
    }
}

/*
 * generate public/private RSA keypairs for all hosts that don't have one.
 */
int
keygen (int bits)
{
  RSA *rsa_key;
  FILE *f;
  char *name = NULL;
  char *fname;

  asprintf (&fname, "%s/hostkeys", confbase);
  mkdir (fname, 0700);
  free (fname);

  asprintf (&fname, "%s/pubkey", confbase);
  mkdir (fname, 0700);
  free (fname);

  for (configuration::node_vector::iterator i = conf.nodes.begin (); i != conf.nodes.end (); ++i)
    {
      conf_node *node = *i;

      asprintf (&fname, "%s/pubkey/%s", confbase, node->nodename);

      f = fopen (fname, "a");

      if (!f)
        {
          perror (fname);
          exit (1);
        }

      if (ftell (f))
        {
          fprintf (stderr, "'%s' already exists, skipping this node\n",
                   fname);
          fclose (f);
          continue;
        }

      fprintf (stderr, _("generating %d bits key for %s:\n"), bits,
               node->nodename);

      rsa_key = RSA_generate_key (bits, 0xFFFF, indicator, NULL);

      if (!rsa_key)
        {
          fprintf (stderr, _("error during key generation!\n"));
          return -1;
        }
      else
        fprintf (stderr, _("Done.\n"));

      PEM_write_RSAPublicKey (f, rsa_key);
      fclose (f);
      free (fname);

      asprintf (&fname, "%s/hostkeys/%s", confbase, node->nodename);

      f = fopen (fname, "a");
      if (!f)
        {
          perror (fname);
          exit (1);
        }

      PEM_write_RSAPrivateKey (f, rsa_key, NULL, NULL, 0, NULL, NULL);
      fclose (f);
      free (fname);
    }

  return 0;
}

int
main (int argc, char **argv, char **envp)
{
  set_identity (argv[0]);
  log_to (LOGTO_STDERR);

  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  parse_options (argc, argv, envp);

  if (show_version)
    {
      printf (_("%s version %s (built %s %s, protocol %d:%d)\n"), get_identity (),
              VERSION, __DATE__, __TIME__, PROTOCOL_MAJOR, PROTOCOL_MINOR);
      printf (_
              ("Copyright (C) 2003 Marc Lehmann <vpe@plan9.de> and others.\n"
               "See the AUTHORS file for a complete list.\n\n"
               "vpe comes with ABSOLUTELY NO WARRANTY.  This is free software,\n"
               "and you are welcome to redistribute it under certain conditions;\n"
               "see the file COPYING for details.\n"));

      return 0;
    }

  if (show_help)
    usage (0);

  make_names ();
  conf.read_config (false);

  if (generate_keys)
    {
      RAND_load_file ("/dev/urandom", 1024);
      exit (keygen (generate_keys));
    }

  if (kill_vped)
    exit (kill_other (kill_vped));

  if (show_config)
    {
      conf.print ();
      exit (0);
    }

  usage (1);
}
